ES6 and multiple shuffled decks

w4dd325 opened the pull request:
- Updated the codebase to utilise ES6 JavaScript, enhancing readability, maintainability, and taking advantage of modern language features.

- Altered the file structure to include a src and test folder.

- Implemented a new function to replicate the deck n times, enabling the concept of 'multiple decks'. The numberOfDecks variable is declared so that the desired number of decks can easily be changed.

- Added a function for shuffling the decks to introduce randomness to the cards. (Ensured that shuffling occurs after deck replication to prevent repetition of the same shuffled decks.)

- Added a summary in the readme.md file to summarise what the project is about.

- In the test.js file, incorporated a loop to visually represent the randomness of cards outputted.

- Additionally, a loop was included to demonstrate the functionality of the slice method, illustrating its ability to remove a card on each iteration until the entire list has been covered.

- Comments were inserted throughout the codebase to aid others.

w4dd325 committed:
ES6 and multiple shuffled decks

Updated the codebase to utilise ES6 JavaScript, enhancing readability, maintainability, and taking advantage of modern language features.

Altered the file structure to include a src and test folder.

Implemented a new function to replicate the deck n times, enabling the concept of 'multiple decks'. The numberOfDecks variable is declared so that the desired number of decks can easily be changed.

Added a function for shuffling the decks to introduce randomness to the cards. (Ensured that shuffling occurs after deck replication to prevent repetition of the same shuffled decks.)

Added a summary in the readme.md file to summarise what the project is about.

In the test.js file, incorporated a loop to visually represent the randomness of cards outputted.

Additionally, a loop was included to demonstrate the functionality of the slice method, illustrating its ability to remove a card on each iteration until the entire list has been covered.

Comments were inserted throughout the codebase to aid others.

dowenb commented on README.md:
```diff
@@ -1,6 +1,10 @@
-A fairer way to provide 2d6 dice rolls
+### A fairer way to provide 2d6 dice rolls!
+
+This GitHub project focuses on enhancing the gaming experience of Settlers of Catan, a board game where players build settlements and cities on a board with numbered tiles representing resources. Players roll 2d6 (2 6-sided dice) to determine which tiles produce resources each turn.
```
Games like settlers of Catan, not only that game. Other similar games would also benefit from this mechanism.

w4dd325 replied:
Change the wording to: 

> This GitHub project focuses on enhancing the gaming experience of games that roll 2 6-sided dice (2d6). For example, Settlers of Catan is a board game where...

dowenb commented on src/index.js:
```diff
@@ -0,0 +1,74 @@
+export default function multiDeck() {
```
By removing/ renaming the original function deck, this will be a breaking change.

w4dd325 replied:
multiDeck renamed back to 'deck'. 
Note:  when importing 'deck' it needs to be assigned to a variable or constant in order to use it. The name cannot be the same as the import name else you get 'SyntaxError: Identifier 'deck' has already been declared'. Therefore, when importing within the test.js file, I have assigned it as follows: const cardDeck = deck();

dowenb commented on src/index.js:
```diff
@@ -0,0 +1,74 @@
+export default function multiDeck() {
+    const dice6 = [1, 2, 3, 4, 5, 6];
+    const numberOfDecks = 5; //<-- This is how many times the deck gets repeated
```
We should have this as a variable you can pass into the function. With 5 as the default.

w4dd325 replied:
Configured the function to use default vaule 5:
export function deck(numberOfDecks = 5) {

And if the user passed another value as an arg it will use that instead, for example: 
const cardDeck = deck(2);

dowenb commented on src/index.js:
```diff
@@ -0,0 +1,74 @@
+    const shuffledRepeatedDeck = shuffleArray(repeatedDeck);
+
+    //Return (output) the shuffled deck
+    return shuffledRepeatedDeck;
```
Shuffle could be separate exported function, used by this function. And could also be optional, incase there is a use case for having them in order.

w4dd325 replied:
Set function as an inline export:
export function shuffleArray(array) {

w4dd325 replied:
I've created a shouldShuffle arg and set the default to true. 

Because there are 2 default args, if only shouldShuffle is passed it wont use the default number of decks as it is expecting 2 args. So I have added this code that checks if only 1 arg is passed, if it is 'not a number' then it sets the variables as needed. 

```
    if (typeof numberOfDecks !== 'number') {
        shouldShuffle = numberOfDecks; // Shift arguments if only one provided
        numberOfDecks = 5; // Set default value for numberOfDecks
    }
```

dowenb commented on src/index.js:
```diff
@@ -0,0 +1,74 @@
+    return repeatedDeck;
+}
+
+// Function to shuffle an array (uses the Fisher-Yates (aka Knuth) Shuffle algorithm.)
```
Oh good this is a function. We could consider exporting this.

w4dd325 replied:
Set function as an inline export:
export function repeatDeck(deck, n) {

dowenb commented on src/index.js:
```diff
@@ -0,0 +1,74 @@
+    }
+    // Return the shuffled array
+    return array;
+}
```
Missing empty line at end of file.

w4dd325 replied:
I have added this in as requested.

dowenb commented on test/test.js:
```diff
@@ -0,0 +1,32 @@
+
+//---------------------------------------------------------------------------------------------------------
+//SPLICING THE DECK
+//The splice() method modifies the original array and returns an array containing the removed element(s).
```
We shouldn't need to explain in comments a standard JavaScript array function.

w4dd325 replied:
Splice comment removed.

w4dd325 committed:
Update version number

Incremented the version number due to breaking changes.

dowenb reviewed (changes requested):
Please rename multiDeck back to deck, and pass the number of decks to generate as a variable, with a default of 5.

Please make the shuffle function optional, you can keep shuffled as the default, but allow it to be left unshuffled.

w4dd325 committed:
PR changes

Changes made as per the requests in the PR.

w4dd325 committed:
Set deck default to 5

Configured the function to use default value of 5. And if a user passes another number as an arg then it will use that value instead.

Added a check to the test.js file that shows both the default deck() and deck(2) function calls working.

w4dd325 committed:
added shouldShuffle

shouldShuffle defaults to true, but a user can override and pass false.

Added some code so that if only 'false' is passed, it will still use the default number of decks.
